File: ConsoleAppProject/(Depth)ConsoleApp.py
#This Python Code Contain pyfiglet
#Also Install pyfiglet lib use pip install pyfiglet in terminal
#This Code Below Is The Code Without Depth if you Dont Like the Depth One You Can Run This Code Below
"""
from math import inf
from random import randint
from pyfiglet import figlet_format
class TicTacToe:
    def __init__(self):
        self.Board = [
            [1,2,3],
            [4,5,6],
            [7,8,9]
        ]
        self.Player = None
        self.AI = None
        self.Diff = None
        self.Turn = None
        self.PrevPlayed = None

    def DrawBoard(self):
        print("-"*13)
        for i in range(3):
           print(f"| {self.Board[i][0]} | {self.Board[i][1]} | {self.Board[i][2]} |")
        print("-"*13)
    
    def ChooseTeam(self):
        while True:
            n = input("Choose You Team (O or X) : ").upper()
            vaildteam = ["O", "X"]
            if n not in vaildteam:
                print("Invaild Team try Again")
                continue
            self.Player = n
            self.AI = "O" if self.Player == "X" else "X"
            break
    
    def ChooseDiff(self):
        while True:
            n = input("Choose You Diff For The AI\ne: Easy\nm: Meduim\nh: Hard\ninput Here : ").lower()
            vaildiff = ['e', 'm', 'h']
            if n not in vaildiff:
                print("Invaild Diff Try Again")
                continue
            self.Diff = n
            break
    def StartTurn(self):
        n = randint(1, 2)
        self.Turn = self.Player if n == 1 else self.AI
    
    def IsTie(self) -> bool:
        return all([type(self.Board[i][j]) != int for i in range(3) for j in range(3)])
    
    def IsGameOver(self) -> bool:
        for i in range(3):
            if self.Board[i][0] == self.Board[i][1] and self.Board[i][1] == self.Board[i][2]:
                return True
            elif self.Board[0][i] == self.Board[1][i] and self.Board[1][i] == self.Board[2][i]:
                return True
        if self.Board[0][0] == self.Board[1][1] and self.Board[1][1] == self.Board[2][2]:
            return True
        elif self.Board[0][2] == self.Board[1][1] and self.Board[1][1] == self.Board[2][0]:
            return True
        return False
    
    def IsAbleToFill(self, row : int, col : int) -> bool:
        return type(self.Board[row][col]) == int
    
    def PlayerFill(self, row, col) -> bool:
        if not self.IsAbleToFill(row, col):
            return
        else:
            self.Board[row][col] = self.Player
    
    def MiniMaxCheckWinner(self):
        for i in range(3):
            if self.Board[i][0] == self.Board[i][1] and self.Board[i][1] == self.Board[i][2]:
                return self.Board[i][0]
            elif self.Board[0][i] == self.Board[1][i] and self.Board[1][i] == self.Board[2][i]:
                return self.Board[0][i]
        if self.Board[0][0] == self.Board[1][1] and self.Board[1][1] == self.Board[2][2]:
            return self.Board[0][0]
        elif self.Board[0][2] == self.Board[1][1] and self.Board[1][1] == self.Board[2][0]:
            return self.Board[0][2]
        return None
    
    def FindAvaSpot(self, symbol : str):
        for i in range(3):
            # Straight Line
            if self.Board[i][0] == symbol and self.Board[i][1] == symbol and self.IsAbleToFill(i, 2):
                return [i, 2]
            elif self.Board[i][2] == symbol and self.Board[i][1] == symbol and self.IsAbleToFill(i, 0):
                return [i, 0]
            elif self.Board[i][2] == symbol and self.Board[i][0] == symbol and self.IsAbleToFill(i, 1):
                return [i, 1]
            #Vertical Line
            elif self.Board[0][i] == symbol and self.Board[1][i] == symbol and self.IsAbleToFill(2, i):
                return [2, i]
            elif self.Board[2][i] == symbol and self.Board[1][i] == symbol and self.IsAbleToFill(0, i):
                return [0, i]
            elif self.Board[2][i] == symbol and self.Board[0][i] == symbol and self.IsAbleToFill(1, i):
                return [1, i]
        # More
        # Side 1
        if self.Board[0][0] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(2, 2):
            return [2, 2]
        elif self.Board[2][2] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(0, 0):
            return [0, 0]
        elif self.Board[2][2] == symbol and self.Board[0][0] == symbol and self.IsAbleToFill(1, 1):
            return [1, 1]
        # Side 2
        elif self.Board[0][2] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(2, 0):
            return [2, 0]
        elif self.Board[2][0] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(0, 2):
            return [0, 2]
        elif self.Board[0][2] == symbol and self.Board[2][0] == symbol and self.IsAbleToFill(1, 1):
            return [1, 1]
        return None
    
    def MiniMax(self, ismaxing, alpha : int , beta : int):
        if self.MiniMaxCheckWinner() == self.AI:
            return 1
        if self.MiniMaxCheckWinner() == self.Player:
            return -1
        if self.IsTie():
            return 0
        # The Algorithm
        if ismaxing:
            # When ismaxing is True
            # AI Do the AI TestCase
            maxscore = -inf
            for i in range(3):
                for j in range(3):
                    if self.IsAbleToFill(i, j):
                        temp = self.Board[i][j]
                        self.Board[i][j] = self.AI
                        score = self.MiniMax(False, alpha, beta)
                        maxscore = max(score, maxscore)
                        alpha = max(alpha, score)
                        self.Board[i][j] = temp
                        if beta <= alpha:
                            break
            return maxscore
        else:
            minscore = inf
            for i in range(3):
                for j in range(3):
                    if self.IsAbleToFill(i, j):
                        temp = self.Board[i][j]
                        self.Board[i][j] = self.Player
                        score = self.MiniMax(True, alpha, beta)
                        minscore = min(score, minscore)
                        beta = min(beta, score)
                        self.Board[i][j] = temp
                        if beta <= alpha:
                            break
            return minscore

    def AIEasyFill(self):
        ava = [[i, j] for i in range(3) for j in range(3) if self.IsAbleToFill(i, j)]
        avaspot = ava[randint(0, len(ava)-1)]
        self.Board[avaspot[0]][avaspot[1]] = self.AI
    
    def AIFill(self):
        match self.Diff:
            case 'e':
                self.AIEasyFill()
            case 'm':
                avaAI = self.FindAvaSpot(self.AI)
                if avaAI:
                    self.Board[avaAI[0]][avaAI[1]] = self.AI
                    return
                avaPlayer = self.FindAvaSpot(self.Player)
                if avaPlayer:
                    self.Board[avaPlayer[0]][avaPlayer[1]] = self.AI
                    return
                # This Code Below Run For The Start of The Game
                for i in range(3):
                    for j in range(3):
                        if self.IsAbleToFill(i, j):
                            self.Board[i][j] = self.AI
                            return
            case "h":
                bestfill = []
                bestscore = -inf
                for i in range(3):
                    for j in range(3):
                        if self.IsAbleToFill(i, j):
                            temp = self.Board[i][j]
                            self.Board[i][j] = self.AI
                            score = self.MiniMax(False, -inf, inf)
                            if score > bestscore:
                                bestscore = score
                                bestfill = [i, j]
                            self.Board[i][j] = temp
                if bestfill:
                    self.Board[bestfill[0]][bestfill[1]] = self.AI

    def AutoPlayRound(self):
        if self.Turn == self.Player:
            while True:
                try:
                    plrcin = int(input(f"Player Turn! Currently Player is : {self.Player}\n Choose Board to Fill in From 1-9\n input Here : "))
                    if plrcin < 1 or plrcin > 9:
                        print("Invalid input, please choose a number between 1 and 9.")
                        continue
                    row, col = (plrcin - 1) // 3, (plrcin - 1) % 3
                    if not self.IsAbleToFill(row, col):
                        print("Board already filled, choose another spot.")
                        continue
                    self.PlayerFill(row, col)
                    break
                except ValueError:
                    print("Invalid input, please enter a number.")
            self.PrevPlayed = self.Player
            self.Turn = self.AI
        else:
            print("AI Turn!..AI Is Thinking And Choosing..")
            self.AIFill()
            self.PrevPlayed = self.AI
            self.Turn = self.Player

    def StartAIGame(self):
        self.ChooseTeam()
        self.ChooseDiff()
        while True:
            self.DrawBoard()
            if self.IsGameOver():
                print("Game Over! The Winner is {}".format(self.PrevPlayed))
                break
            if self.IsTie():
                print("Game Tie!")
                break
            self.AutoPlayRound()
                
# Test Result
if __name__ == "__main__": 
    print(figlet_format("TicTacToe..."))
    ttt1 = TicTacToe()
    ttt1.StartAIGame()
"""
#This Code Below is The Code With Depth Algorithm I dont use Node Since Is Doesnt matter Cus My Code Working Fine
from math import inf
from random import randint
from pyfiglet import figlet_format
class TicTacToe:
    def __init__(self):
        self.Board = [
            [1,2,3],
            [4,5,6],
            [7,8,9]
        ]
        self.Player = None
        self.AI = None
        self.Diff = None
        self.Turn = None
        self.PrevPlayed = None

    def DrawBoard(self):
        print("-"*13)
        for i in range(3):
           print(f"| {self.Board[i][0]} | {self.Board[i][1]} | {self.Board[i][2]} |")
        print("-"*13)
    
    def ChooseTeam(self):
        while True:
            n = input("Choose You Team (O or X) : ").upper()
            vaildteam = ["O", "X"]
            if n not in vaildteam:
                print("Invaild Team try Again")
                continue
            self.Player = n
            self.AI = "O" if self.Player == "X" else "X"
            break
    
    def ChooseDiff(self):
        while True:
            n = input("Choose You Diff For The AI\ne: Easy\nm: Meduim\nh: Hard\ninput Here : ").lower()
            vaildiff = ['e', 'm', 'h']
            if n not in vaildiff:
                print("Invaild Diff Try Again")
                continue
            self.Diff = n
            break
    def StartTurn(self):
        n = randint(1, 2)
        self.Turn = self.Player if n == 1 else self.AI
    
    def IsTie(self) -> bool:
        return all([type(self.Board[i][j]) != int for i in range(3) for j in range(3)])
    
    def IsGameOver(self) -> bool:
        for i in range(3):
            if self.Board[i][0] == self.Board[i][1] and self.Board[i][1] == self.Board[i][2]:
                return True
            elif self.Board[0][i] == self.Board[1][i] and self.Board[1][i] == self.Board[2][i]:
                return True
        if self.Board[0][0] == self.Board[1][1] and self.Board[1][1] == self.Board[2][2]:
            return True
        elif self.Board[0][2] == self.Board[1][1] and self.Board[1][1] == self.Board[2][0]:
            return True
        return False
    
    def IsAbleToFill(self, row : int, col : int) -> bool:
        return type(self.Board[row][col]) == int
    
    def PlayerFill(self, row, col) -> bool:
        if not self.IsAbleToFill(row, col):
            return
        else:
            self.Board[row][col] = self.Player
    
    def MiniMaxCheckWinner(self):
        for i in range(3):
            if self.Board[i][0] == self.Board[i][1] and self.Board[i][1] == self.Board[i][2]:
                return self.Board[i][0]
            elif self.Board[0][i] == self.Board[1][i] and self.Board[1][i] == self.Board[2][i]:
                return self.Board[0][i]
        if self.Board[0][0] == self.Board[1][1] and self.Board[1][1] == self.Board[2][2]:
            return self.Board[0][0]
        elif self.Board[0][2] == self.Board[1][1] and self.Board[1][1] == self.Board[2][0]:
            return self.Board[0][2]
        return None
    
    def FindAvaSpot(self, symbol : str):
        for i in range(3):
            # Straight Line
            if self.Board[i][0] == symbol and self.Board[i][1] == symbol and self.IsAbleToFill(i, 2):
                return [i, 2]
            elif self.Board[i][2] == symbol and self.Board[i][1] == symbol and self.IsAbleToFill(i, 0):
                return [i, 0]
            elif self.Board[i][2] == symbol and self.Board[i][0] == symbol and self.IsAbleToFill(i, 1):
                return [i, 1]
            #Vertical Line
            elif self.Board[0][i] == symbol and self.Board[1][i] == symbol and self.IsAbleToFill(2, i):
                return [2, i]
            elif self.Board[2][i] == symbol and self.Board[1][i] == symbol and self.IsAbleToFill(0, i):
                return [0, i]
            elif self.Board[2][i] == symbol and self.Board[0][i] == symbol and self.IsAbleToFill(1, i):
                return [1, i]
        # More
        # Side 1
        if self.Board[0][0] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(2, 2):
            return [2, 2]
        elif self.Board[2][2] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(0, 0):
            return [0, 0]
        elif self.Board[2][2] == symbol and self.Board[0][0] == symbol and self.IsAbleToFill(1, 1):
            return [1, 1]
        # Side 2
        elif self.Board[0][2] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(2, 0):
            return [2, 0]
        elif self.Board[2][0] == symbol and self.Board[1][1] == symbol and self.IsAbleToFill(0, 2):
            return [0, 2]
        elif self.Board[0][2] == symbol and self.Board[2][0] == symbol and self.IsAbleToFill(1, 1):
            return [1, 1]
        return None
    
    def MiniMax(self, ismaxing, alpha : int , beta : int, depth : int):
        winner = self.MiniMaxCheckWinner()
        if winner == self.AI:
            return 1 - depth
        if winner == self.Player:
            return -1 + depth
        if self.IsTie():
            return 0
        # The Algorithm
        if ismaxing:
            # When ismaxing is True
            # AI Do the AI TestCase
            maxscore = -inf
            for i in range(3):
                for j in range(3):
                    if self.IsAbleToFill(i, j):
                        temp = self.Board[i][j]
                        self.Board[i][j] = self.AI
                        score = self.MiniMax(False, alpha, beta, depth+1)
                        maxscore = max(score, maxscore)
                        alpha = max(alpha, score)
                        self.Board[i][j] = temp
                        if beta <= alpha:
                            break
            return maxscore
        else:
            minscore = inf
            for i in range(3):
                for j in range(3):
                    if self.IsAbleToFill(i, j):
                        temp = self.Board[i][j]
                        self.Board[i][j] = self.Player
                        score = self.MiniMax(True, alpha, beta, depth + 1)
                        minscore = min(score, minscore)
                        beta = min(beta, score)
                        self.Board[i][j] = temp
                        if beta <= alpha:
                            break
            return minscore

    def AIEasyFill(self):
        ava = [[i, j] for i in range(3) for j in range(3) if self.IsAbleToFill(i, j)]
        avaspot = ava[randint(0, len(ava)-1)]
        self.Board[avaspot[0]][avaspot[1]] = self.AI
    
    def AIFill(self):
        match self.Diff:
            case 'e':
                self.AIEasyFill()
            case 'm':
                avaAI = self.FindAvaSpot(self.AI)
                if avaAI:
                    self.Board[avaAI[0]][avaAI[1]] = self.AI
                    return
                avaPlayer = self.FindAvaSpot(self.Player)
                if avaPlayer:
                    self.Board[avaPlayer[0]][avaPlayer[1]] = self.AI
                    return
                # This Code Below Run For The Start of The Game
                for i in range(3):
                    for j in range(3):
                        if self.IsAbleToFill(i, j):
                            self.Board[i][j] = self.AI
                            return
            case "h":
                bestfill = []
                bestscore = -inf
                for i in range(3):
                    for j in range(3):
                        if self.IsAbleToFill(i, j):
                            temp = self.Board[i][j]
                            self.Board[i][j] = self.AI
                            score = self.MiniMax(False, -inf, inf, 0)
                            if score > bestscore:
                                bestscore = score
                                bestfill = [i, j]
                            self.Board[i][j] = temp
                if bestfill:
                    self.Board[bestfill[0]][bestfill[1]] = self.AI

    def AutoPlayRound(self):
        if self.Turn == self.Player:
            while True:
                try:
                    plrcin = int(input(f"Player Turn! Currently Player is : {self.Player}\n Choose Board to Fill in From 1-9\n input Here : "))
                    if plrcin < 1 or plrcin > 9:
                        print("Invalid input, please choose a number between 1 and 9.")
                        continue
                    row, col = (plrcin - 1) // 3, (plrcin - 1) % 3
                    if not self.IsAbleToFill(row, col):
                        print("Board already filled, choose another spot.")
                        continue
                    self.PlayerFill(row, col)
                    break
                except ValueError:
                    print("Invalid input, please enter a number.")
            self.PrevPlayed = self.Player
            self.Turn = self.AI
        else:
            print("AI Turn!..AI Is Thinking And Choosing..")
            self.AIFill()
            self.PrevPlayed = self.AI
            self.Turn = self.Player

    def StartAIGame(self):
        self.ChooseTeam()
        self.ChooseDiff()
        while True:
            self.DrawBoard()
            if self.IsGameOver():
                print("Game Over! The Winner is {}".format(self.PrevPlayed))
                break
            if self.IsTie():
                print("Game Tie!")
                break
            self.AutoPlayRound()
                
# Test Result
if __name__ == "__main__": 
    print(figlet_format("TicTacToe..."))
    ttt1 = TicTacToe()
    ttt1.StartAIGame()

File: ConsoleAppProject/(Depth)ConsoleApp.lua
local tictactoe = {}
tictactoe.__index = tictactoe

function tictactoe.new()
    local self = setmetatable({}, tictactoe)
    self.Board = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    }
    self.Player = nil
    self.AI = nil
    self.Diff = nil
    self.PrevPlayed = nil
    self.Turn = nil
    return self
end

-- Lua Index of Table Start With 1
-- So I use 1 as First Index
function tictactoe:DrawBoard()
    print(string.rep("-", 14))
    for i = 1, 3 do
        print("|"..tostring(self.Board[i][1]).." | "..tostring(self.Board[i][2]).." | "..tostring(self.Board[i][3]).. " |")
    end
    print(string.rep("-", 14))
end
function tictactoe:ChooseTeam()
    ::again::
    io.write("Choose You Team (O or X ) : ")
    local n = string.upper(io.read())
    if n == "O" then
        self.Player = n
        self.AI = "X"
    elseif n == "X" then
        self.Player = n
        self.AI = "O"
    else
        print("Invaild Team")
        goto again
    end
end

function tictactoe:SelectDiff()
    ::again::
    io.write("Choose You AI Diff\ne: Easy\nm: Meduim\nh: Hard\nInput Here : ")
    local n = string.lower(io.read())
    if n == "e" then
        self.Diff = n
    elseif n == "m" then
        self.Diff = n
    elseif n == "h" then
        self.Diff = n
    else
        print("Invaild Input")
        goto again
    end
end

function tictactoe:StartTurn()
    local n = math.random(1, 2)
    if n == 1 then
        self.Turn = self.Player
    else
        self.Turn = self.AI
    end
end

function tictactoe:IsAbleToFill(row, col)
    return type(self.Board[row][col]) == "number"
end

function tictactoe:IsTie()
    for i = 1, 3 do
        for j = 1, 3 do
            if self:IsAbleToFill(i, j) then
                return false
            end
        end
    end
    return true
end

function tictactoe:IsGameOver()
    for i = 1, 3 do
        if self.Board[i][1] == self.Board[i][2] and self.Board[i][2] == self.Board[i][3] then
            return true
        elseif self.Board[1][i] == self.Board[2][i] and self.Board[2][i] == self.Board[3][i] then
            return true
        end
    end
    if self.Board[1][1] == self.Board[2][2] and self.Board[2][2] == self.Board[3][3] then
        return true
    elseif self.Board[1][3] == self.Board[2][2] and self.Board[2][2] == self.Board[3][1] then
        return true
    end
end

function tictactoe:PlayerFill(row, col)
    if self:IsAbleToFill(row, col) then
        self.Board[row][col] = self.Player
    else
        return
    end
end

-- Might Confunse You on This One
-- Function For AI (MiniMax Algorithm/AI Algorithm)
function tictactoe:EasyAIFill()
    local ava = {}
    for i = 1, 3 do
        for j = 1, 3 do
            if self:IsAbleToFill(i, j) then
                table.insert(ava, {i, j})
            end
        end
    end
    if ava then
        local avaSpot = ava[math.random(1, #ava)]
        self.Board[avaSpot[1]][avaSpot[2]] = self.AI
    end
end

function tictactoe:FindAvaSpot(symbol)
    for i= 1,  3 do
        if self.Board[i][1] == symbol and self.Board[i][2] == symbol and self:IsAbleToFill(i, 3) then
            return {i, 3}
        elseif self.Board[i][3] == symbol and self.Board[i][2] == symbol and self:IsAbleToFill(i, 1) then
            return {i, 1}
        elseif self.Board[1][i] == symbol and self.Board[2][i] == symbol and self:IsAbleToFill(3, i) then
            return {3, i}
        elseif self.Board[3][i] == symbol and self.Board[2][i] == symbol and self:IsAbleToFill(1, i) then
            return {1, i}
        elseif self.Board[3][i] == symbol and self.Board[1][i] == symbol and self:IsAbleToFill(2, i) then
            return {2, i}
        elseif self.Board[i][3] == symbol and self.Board[i][1] == symbol and self:IsAbleToFill(i, 2) then
            return {i, 2}
        end
    end
    -- Ava Spot 2 For AI Meduim
    if self.Board[1][1] == symbol and self.Board[2][2] == symbol and self:IsAbleToFill(3, 3) then return {3, 3} end
    if self.Board[3][3] == symbol and self.Board[2][2] == symbol and self:IsAbleToFill(1, 1) then return {1, 1} end
    if self.Board[1][3] == symbol and self.Board[2][2] == symbol and self:IsAbleToFill(3, 1) then return {3, 1} end
    if self.Board[3][1] == symbol and self.Board[2][2] == symbol and self:IsAbleToFill(1, 3) then return {1, 3} end
    return nil
end
-- Now This One Gonna Be The Main Algorithm for The Minimax AI which is a Hard Mode AI
-- This One For Checking Winner With MiniMax Or Maybe you can use The IsGameOver() Func Which i Made this Optional
-- Well is Just Waste of Line You Can Just Replace This With IsGameOver And Make the IsGameOver Reutrn char instead of Bool
function tictactoe:MiniMaxCheckWinner()
    for i = 1, 3 do
        if self.Board[i][1] == self.Board[i][2] and self.Board[i][2] == self.Board[i][3] then
            return self.Board[i][1]
        elseif self.Board[1][i] == self.Board[2][i] and self.Board[2][i] == self.Board[3][i] then
            return self.Board[1][i]
        end
    end
    if self.Board[1][1] == self.Board[2][2] and self.Board[2][2] == self.Board[3][3] then
        return self.Board[1][1]
    elseif self.Board[1][3] == self.Board[2][2] and self.Board[2][2] == self.Board[3][1] then
        return self.Board[1][3]
    end
    return nil
end
-- Now All We Need is Just This Function For The Hard AI Algorithm
function tictactoe:MiniMax(ismaxing, alpha, beta, depth)
    local winner = self:MiniMaxCheckWinner()
    if winner == self.AI then return 1 - depth end
    if winner == self.Player then return -1 + depth end
    if self:IsTie() then return 0 end
    if ismaxing then
        -- This One For The AI Case
        -- Such That AI Going Check All Possible Way to Win The Player With This Algorithm
        local maxscore = -math.huge
        for i = 1, 3 do
            for j = 1, 3 do
                if self:IsAbleToFill(i, j) then
                    local temp = self.Board[i][j] 
                    self.Board[i][j] = self.AI
                    local score = self:MiniMax(false, alpha, beta, depth + 1)
                    maxscore = math.max(score, maxscore)
                    self.Board[i][j] = temp
                    alpha = math.max(alpha, score)
                    if beta <= alpha then
                        break
                    end

                end
            end
        end
        return maxscore
    else
        -- This One For The Player Case
        -- Such That AI Going To Play The Player To Find That How Will AI Win or Block Player Spot When Player is Filling(Might not Accruate)
        local minscore = math.huge
        for i = 1, 3 do
            for j = 1, 3 do
                if self:IsAbleToFill(i, j) then
                    local temp = self.Board[i][j]
                    self.Board[i][j] = self.Player
                    local score = self:MiniMax(true, alpha, beta, depth + 1)
                    minscore = math.min(score, minscore)
                    self.Board[i][j] = temp
                    beta = math.min(beta, score) 
                    if beta <= alpha then break end

                end
            end
        end
        return minscore
    end
end

function tictactoe:AIFill()
    if self.Diff == "e" then
        self:EasyAIFill()
    elseif self.Diff == "m" then
        local WinSpot = self:FindAvaSpot(self.AI)
        if WinSpot then
            self.Board[WinSpot[1]][WinSpot[2]] = self.AI
            return
        end
        local BlockSpot = self:FindAvaSpot(self.Player) 
        if BlockSpot then
            self.Board[BlockSpot[1]][BlockSpot[2]] = self.AI
            return
        end
        -- This Code Run When There No WinSpot or BlockSpot
        for i = 1, 3 do
            for j = 1, 3 do
                if self:IsAbleToFill(i, j) then
                    self.Board[i][j] = self.AI
                    return
                end
            end
        end
    elseif self.Diff == "h" then
        -- This One Use The Algorithm For Winning So Feel Free To View The Soucre of the Algorithm
        local bestscore = -math.huge
        local bestfill = nil
        for i = 1, 3 do
            for j = 1, 3 do
                if self:IsAbleToFill(i, j) then
                    local temp = self.Board[i][j] 
                    self.Board[i][j] = self.AI
                    local score = self:MiniMax(false, -math.huge, math.huge, 0)
                    if score > bestscore then
                        bestscore = score
                        bestfill = {i, j}
                    end
                    self.Board[i][j] = temp
                end
            end
        end
        if bestfill then
            self.Board[bestfill[1]][bestfill[2]] = self.AI
        end
    end
end

-- Function For Main Game Code
function tictactoe:AutoPlayRound()
    if self.Turn == self.Player then
        -- Player's turn
        local validInput = false
        while not validInput do
            io.write(string.format("Player Turn! Currently Player Is %s\nChoose To Fill From 1-9\ninput Here : ", self.Player))
            local plrCin = tonumber(io.read())
            if plrCin and plrCin >= 1 and plrCin <= 9 then
                local row = math.ceil(plrCin / 3)
                local col = (plrCin - 1) % 3 + 1
                if self:IsAbleToFill(row, col) then
                    self:PlayerFill(row, col)
                    validInput = true
                else
                    print("Board Already Filled. Choose Another.")
                end
            else
                print("Invalid input. Please enter a number between 1 and 9.")
            end
        end
        self.PrevPlayed = self.Player
        self.Turn = self.AI
    else
        -- AI's turn
        print("AI Turn!.. AI is Thinking.. and Choosing")
        self:AIFill()
        self.PrevPlayed = self.AI
        self.Turn = self.Player
    end
end

-- Fit In All Function Together into Simple Code
function tictactoe:StartAIGame()
    self:ChooseTeam()
    self:SelectDiff()
    self:StartTurn()
    while true do
        --os.execute("clear")
        self:DrawBoard()
        if self:IsGameOver() then print("Game Over Winner : " .. self.PrevPlayed) break end
        if self:IsTie() then print("Game Tie!") break end
        self:AutoPlayRound()
    end
end

-- Test Result
local ttt1 = tictactoe.new()
ttt1:StartAIGame()
